tzrec/models/masknet.py

```diff
@@ -0,0 +1,161 @@
+    """MaskBlock module.
+
+    Args:
+        input_dim (int): Input dimension, either feature embedding dim(parallel mode)
+            or hidden state dim(serial mode).
+        mask_input_dim (int): Mask input dimension, is always the feature embedding dim
+            for both para and serial modes.
+        reduction_ratio (float): Reduction ratio, aggregation_dim / mask_input_dim.
+        aggregation_dim (int): Aggregation layer dim, mask_input_dim*reduction_ratio.
+        hidden_dim (int): Hidden layer dimension for feedforward network.
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        mask_input_dim: int,
+        reduction_ratio: float,
+        aggregation_dim: int,
+        hidden_dim: int,
+    ) -> None:
+        super(MaskBlock, self).__init__()
+        self.ln_emb = nn.LayerNorm(input_dim)
+
+        if not aggregation_dim and not reduction_ratio:
+            raise ValueError(
+                "Either aggregation_dim or reduction_ratio must be provided."
+            )
+
+        if aggregation_dim:
+            self.aggregation_dim = aggregation_dim
+        if reduction_ratio:
+            self.aggregation_dim = int(mask_input_dim * reduction_ratio)
+
+        assert self.aggregation_dim > 0, (
+            "aggregation_dim must be > 0, check your aggregation_dim or "
+        )
+        "redudction_ratio settings."
+
+        self.mask_generator = nn.Sequential(
+            nn.Linear(mask_input_dim, self.aggregation_dim),
+            nn.ReLU(),
+            nn.Linear(self.aggregation_dim, input_dim),
+        )
+
+        assert hidden_dim > 0, "hidden_dim must be > 0."
+
+        self.ffn = nn.Sequential(
+            nn.Linear(input_dim, hidden_dim),
+            nn.LayerNorm(hidden_dim),
+            nn.ReLU(),
+            nn.Linear(hidden_dim, input_dim),
+        )
+
+    def forward(self, input: torch.Tensor, mask_input: torch.Tensor) -> torch.Tensor:
+        """Forward pass of MaskBlock."""
+        ln_emb = self.ln_emb(input)
+        weights = self.mask_generator(mask_input)
+        weighted_emb = ln_emb * weights
+        output = self.ffn(weighted_emb)
+
+        return output
+
+
+class MaskNet(RankModel):
+    """Masknet model.
+
+    Args:
+        model_config (ModelConfig): an instance of ModelConfig.
+        features (list): list of features.
+        labels (list): list of label names.
+        sample_weights (list): sample weight names.
+    """
+
+    def __init__(
+        self,
+        model_config: ModelConfig,
+        features: List[BaseFeature],
+        labels: List[str],
+        sample_weights: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(model_config, features, labels, sample_weights, **kwargs)
+        self.init_input()
+        self.group_name = self.embedding_group.group_names()[0]
+        feature_dim = self.embedding_group.group_total_dim(self.group_name)
+
+        masknet_config = model_config.mask_net
```

Comment:
The logic of mask_net is suggested to be moved into a module in modules directory to facilitate its use as a submodule